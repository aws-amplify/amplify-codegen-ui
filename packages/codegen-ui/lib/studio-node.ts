/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { StudioComponent, StudioComponentChild } from './types';

export class StudioNode {
  component: StudioComponent | StudioComponentChild;

  parent?: StudioNode;

  constructor(component: StudioComponent | StudioComponentChild, parent?: StudioNode) {
    this.component = component;
    this.parent = parent;
  }

  isRoot(): boolean {
    return this.parent === undefined;
  }

  getComponentPathToRoot(): StudioNode[] {
    if (this.parent !== undefined) {
      return [this as StudioNode].concat(this.parent.getComponentPathToRoot());
    }
    return [this];
  }

  /**
   * Return true if any ancestor node of the current node has type `componentType`.
   */
  hasAncestorOfType(componentType: string): boolean {
    const ancestorComponentTypes = this.getComponentPathToRoot().map((node) => node.component.componentType);
    // We don't want to check if the current node has type, so shift that element out.
    ancestorComponentTypes.shift();
    return ancestorComponentTypes.some((ancestorComponentType) => ancestorComponentType === componentType);
  }
}
