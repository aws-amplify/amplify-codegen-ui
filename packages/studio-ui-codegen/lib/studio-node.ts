/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { StudioComponent, StudioComponentChild } from './types';

export class StudioNode {
  component: StudioComponent | StudioComponentChild;

  parent?: StudioNode;

  children: StudioNode[];

  elementIndex?: number;

  constructor(component: StudioComponent | StudioComponentChild, parent?: StudioNode) {
    this.component = component;
    this.parent = parent;
    this.children = [];
    if (this.parent) {
      this.elementIndex = this.parent.children.filter(
        (childNode) => this.component.componentType === childNode.component.componentType,
      ).length;
      this.parent.children.push(this);
    }
  }

  isRoot(): boolean {
    return this.parent === undefined;
  }

  getComponentPathToRoot(): StudioNode[] {
    if (this.parent !== undefined) {
      return [this as StudioNode].concat(this.parent.getComponentPathToRoot());
    }
    return [this];
  }

  /**
   * Build the override path for a given element walking from the node to tree root, providing an index
   * for all but the top-level components.
   * Example:
   * <Flex> <-- returns 'Flex'
   *     <Button> <-- returns 'Flex.Button[0]'
   *     <Button> <-- returns 'Flex.Button[1]'
   *     <Flex> <-- returns 'Flex.Flex[0]'
   *         </Button> <-- returns 'Flex.Flex[0].Button[0]'
   *     </Flex>
   * </Flex>
   */
  getOverrideKey(): string {
    const [parentElement, ...childElements] = this.getComponentPathToRoot().reverse();
    const childPath = childElements.map((node) => `${node.component.componentType}[${node.elementIndex}]`);
    return [parentElement.component.componentType, ...childPath].join('.');
  }
}
