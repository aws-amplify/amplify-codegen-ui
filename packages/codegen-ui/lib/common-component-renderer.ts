/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { StudioComponent, StudioComponentChild, WrappedComponentProperties } from './types';
import { StudioNode } from './studio-node';
import { validateComponentSchema } from './validation-helper';

/**
 * Shared class for rendering components.
 * Mostly contains helper functions for mapping the Studio schema to actual props.
 */
export abstract class CommonComponentRenderer<TPropIn> {
  protected inputProps: WrappedComponentProperties<TPropIn>;

  protected node: StudioNode;

  constructor(protected component: StudioComponent | StudioComponentChild, protected parent?: StudioNode) {
    // Run schema validation on the top-level component.
    if (this.parent === undefined) {
      validateComponentSchema(this.component as StudioComponent);
    }
    const flattenedProps = Object.entries(component.properties).map((prop) => {
      return [prop[0], prop[1]];
    });
    this.inputProps = Object.fromEntries(flattenedProps);
    this.node = new StudioNode(component, parent);
  }
}
