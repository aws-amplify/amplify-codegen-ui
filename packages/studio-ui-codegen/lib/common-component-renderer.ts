import {
  FixedStudioComponentProperty,
  StudioComponent,
  StudioComponentChild,
  StudioComponentProperties,
  WrappedComponentProperties,
} from '@amzn/amplify-ui-codegen-schema';

import { StudioNode } from './studio-node';

/**
 * Shared class for rendering components.
 * Mostly contains helper functions for mapping the Studio schema to actual props.
 */
export abstract class CommonComponentRenderer<TPropIn> {
  protected inputProps: WrappedComponentProperties<TPropIn>;

  protected node: StudioNode;

  constructor(protected component: StudioComponent | StudioComponentChild, protected parent?: StudioNode) {
    const flattenedProps = Object.entries(component.properties).map((prop) => {
      return [prop[0], prop[1]];
    });
    this.inputProps = Object.fromEntries(flattenedProps);
    this.node = new StudioNode(component, parent);
  }
}
