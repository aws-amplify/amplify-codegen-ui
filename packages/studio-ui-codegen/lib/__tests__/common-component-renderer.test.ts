import { BaseComponentProps } from '@aws-amplify/ui-react';
import { StudioNode } from '../studio-node';
import { CommonComponentRenderer } from '../common-component-renderer';

class MockComponentRenderer extends CommonComponentRenderer<BaseComponentProps> {}

describe('common-component-renderer', () => {
  test('constructor', () => {
    const component = {
      componentType: 'Button',
      name: 'MyButton',
      properties: {
        value: { value: 'Confirm' },
      },
    };
    const parent = new StudioNode({
      componentType: 'View',
      name: 'MyView',
      properties: {},
    });
    const renderer = new MockComponentRenderer(component, parent);
    expect(renderer).toMatchSnapshot();
  });
});
