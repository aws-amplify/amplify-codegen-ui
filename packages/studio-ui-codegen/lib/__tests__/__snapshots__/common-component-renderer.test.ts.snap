// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common-component-renderer constructor 1`] = `
MockComponentRenderer {
  "component": Object {
    "componentType": "Button",
    "name": "MyButton",
    "properties": Object {
      "value": Object {
        "value": "Confirm",
      },
    },
  },
  "inputProps": Object {
    "value": Object {
      "value": "Confirm",
    },
  },
  "node": StudioNode {
    "children": Array [],
    "component": Object {
      "componentType": "Button",
      "name": "MyButton",
      "properties": Object {
        "value": Object {
          "value": "Confirm",
        },
      },
    },
    "elementIndex": 0,
    "parent": StudioNode {
      "children": Array [
        [Circular],
      ],
      "component": Object {
        "componentType": "View",
        "name": "MyView",
        "properties": Object {},
      },
      "parent": undefined,
    },
  },
  "parent": StudioNode {
    "children": Array [
      StudioNode {
        "children": Array [],
        "component": Object {
          "componentType": "Button",
          "name": "MyButton",
          "properties": Object {
            "value": Object {
              "value": "Confirm",
            },
          },
        },
        "elementIndex": 0,
        "parent": [Circular],
      },
    ],
    "component": Object {
      "componentType": "View",
      "name": "MyView",
      "properties": Object {},
    },
    "parent": undefined,
  },
}
`;
